File: 1140-stone-game-ii/1140-stone-game-ii.py
class Solution:
    def stoneGameII(self, piles: List[int]) -> int:
        # dp array for memo
        dp = [[[-1 for _ in range(101)] for _ in range(101)] for _ in range(2)]

        n = len(piles)

        def solve(person, index, m):
            # if index reached so alice no stones left :(
            if index >= n:
                return 0

            # check if common subproblem here
            if dp[person][index][m] != -1:
                return dp[person][index][m]


            #as we compute max for alice and min for BOB
            result = -1 if person == 1 else float("inf")

            stones = 0

            for x in range(1, min(2 * m, n - index) + 1):

                #add current stones 
                stones += piles[index + x - 1]

                if person == 1:  #if Alice
                    result = max(result, stones + solve(0, index + x, max(m, x)))

                else: #if Bob
                    result = min(result, solve(1, index + x, max(m, x)))

            dp[person][index][m] = result #store result in dp
            
            return result

        return solve(1, 0, 1)

File: 1140-stone-game-ii/1140-stone-game-ii.cpp
class Solution {
public:
    int n;
    int dp[2][101][101];
    int solve(int person , int index , int m , vector<int> &piles){
        if (index >= n) return 0;

        if (dp[person][index][m] != -1) return dp[person][index][m];

        int result;

        if (person == 1) result = -1;
        else result=INT_MAX;

        int stones = 0;

        for (int x = 1 ; x <= min(2*m , n-index) ; x++){
            
            stones+= piles[index + x - 1];

            if (person == 1){ //Alice
                result = max(result , stones + solve(0 , index+x , max(m , x) , piles));
            }

            else{ //BOB
                result = min(result , solve(1 , index+x , max(m , x) , piles));
            }


        }

        return  dp[person][index][m] = result;
    }

    int stoneGameII(vector<int>& piles) {
        n = piles.size();
        memset (dp , -1 , sizeof(dp));
        return solve(1 , 0 , 1 , piles);
    }
};